File: Csla20Test/Csla.Test/ValidationRules/HasRulesManager.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Csla.Test.ValidationRules
{
    [Serializable]
    public class HasRulesManager : BusinessBase<HasRulesManager>
    {
        private string _name;

        protected override object GetIdValue()
        {
            return _name;
        }

        public string Name
        {
            get { return _name; }
            set
            {
                if (_name != value)
                    _name = value;
                ValidationRules.CheckRules("Name");
                MarkDirty();
            }
        }

        protected override void AddBusinessRules()
        {
            ValidationRules.AddRule(NameRequired, "Name");
            ValidationRules.AddRule(NameLength, new MaxLengthArgs("Name", 10));
            ValidationRules.CheckRules();
        }

        private bool NameRequired(object target, Validation.RuleArgs e)
        {
            if (_name.Length > 0)
            {
                return true;
            }
            else
            {
                e.Description = "Name required";
                return false;
            }
        }

        private bool NameLength(object target, Validation.RuleArgs e)
        {
            if (_name.Length <= ((MaxLengthArgs)(e)).MaxLength)
            {
                return true;
            }
            else
            {
                e.Description = "The value for " + e.PropertyName + " is too long";
                return false;
            }
        }

        [Serializable]
        private class Criteria
        {
            public string _name;

            public Criteria()
            {
                _name = "<new>";
            }

            public Criteria(string name)
            {
                this._name = name;
            }
        }

        public static HasRulesManager NewHasRulesManager()
        {
            return (HasRulesManager)(Csla.DataPortal.Create(new Criteria()));
        }

        public static HasRulesManager GetHasRulesManager(string name)
        {
            return (HasRulesManager)(Csla.DataPortal.Fetch(new Criteria(name)));
        }

        public static void DeleteHasRulesManager(string name)
        {
            Csla.DataPortal.Delete(new Criteria(name));
        }

        private HasRulesManager()
        {
            //prevent direct creation
            AddBusinessRules();
        }

        protected override void DataPortal_Create(object criteria)
        {
            Criteria crit = (Criteria)(criteria);
            _name = crit._name;
            Csla.ApplicationContext.GlobalContext.Add("HasRulesManager", "Created");
        }

        protected override void DataPortal_Fetch(object criteria)
        {
            Criteria crit = (Criteria)(criteria);
            _name = crit._name;
            MarkOld();
            Csla.ApplicationContext.GlobalContext.Add("HasRulesManager", "Fetched");
        }

        protected override void DataPortal_Update()
        {
            if (IsDeleted)
            {
                //we would delete here
                Csla.ApplicationContext.GlobalContext.Add("HasRulesManager", "Fetched");
                MarkNew();
            }
            else
            {
                if (IsNew)
                {
                    //we would insert here
                    Csla.ApplicationContext.GlobalContext.Add("HasRulesManager", "Inserted");
                }
                else
                {
                    //we would update here
                    Csla.ApplicationContext.GlobalContext.Add("HasRulesManager", "Updated");
                }
                MarkOld();
            }
        }

        protected override void DataPortal_Delete(object criteria)
        {
            //we would delete here
            Csla.ApplicationContext.GlobalContext.Add("HasRulesManager", "Deleted");
        }


    }
}

File: CSLA10/CSLA.Test/HasRulesManager.vb
Imports CSLA.BrokenRules

<Serializable()> _
Public Class HasRulesManager
  Inherits BusinessBase

  Private mName As String = ""

  Public Property Name() As String
    Get
      Return mName
    End Get
    Set(ByVal Value As String)
      If mName <> Value Then
        mName = Value
        BrokenRules.CheckRules("Name")
        MarkDirty()
      End If
    End Set
  End Property

  Protected Overrides Sub AddBusinessRules()
    With BrokenRules
      .SetTargetObject(Me)
      .AddRule(AddressOf NameRequired, "Name")
      .AddRule(AddressOf NameLength, "Name", New MaxLengthArgs(10))
      .CheckRules()
    End With
  End Sub

  <Description("{0} required")> _
  Private Function NameRequired(ByVal target As Object, ByVal e As RuleArgs) As Boolean
    Return Len(mName) > 0
  End Function

  <Description("{0} too long")> _
  Private Function NameLength(ByVal target As Object, ByVal e As RuleArgs) As Boolean
    Return Len(mName) <= DirectCast(e, MaxLengthArgs).MaxLength
  End Function

  <Serializable()> _
  Private Class Criteria
    Public Name As String

    Public Sub New()
      Name = "<new>"
    End Sub

    Public Sub New(ByVal Name As String)
      Me.Name = Name
    End Sub
  End Class

  Public Shared Function NewHasRulesManager() As HasRulesManager
    Return DirectCast(DataPortal.Create(New Criteria()), HasRulesManager)
  End Function

  Public Shared Function GetHasRulesManager(ByVal Name As String) As HasRulesManager
    Return DirectCast(DataPortal.Fetch(New Criteria(Name)), HasRulesManager)
  End Function

  Public Shared Sub DeleteHasRulesManager(ByVal Name As String)
    DataPortal.Delete(New Criteria(Name))
  End Sub


  Private Sub New()
    ' prevent direct creation
    AddBusinessRules()
  End Sub

  Protected Overrides Sub DataPortal_Create(ByVal Criteria As Object)
    Dim crit As Criteria = DirectCast(Criteria, Criteria)
    mName = crit.Name
    Session.Add("HasRulesManager", "Created")
  End Sub

  Protected Overrides Sub DataPortal_Fetch(ByVal Criteria As Object)
    Dim crit As Criteria = DirectCast(Criteria, Criteria)
    mName = crit.Name
    MarkOld()
    Session.Add("HasRulesManager", "Fetched")
  End Sub

  Protected Overrides Sub DataPortal_Update()
    If IsDeleted Then
      ' we would delete here
      Session.Add("HasRulesManager", "Deleted")
      MarkNew()
    Else
      If IsNew Then
        ' we would insert here
        Session.Add("HasRulesManager", "Inserted")

      Else
        ' we would update here
        Session.Add("HasRulesManager", "Updated")
      End If
      MarkOld()
    End If
  End Sub

  Protected Overrides Sub DataPortal_Delete(ByVal Criteria As Object)
    ' we would delete here
    Session.Add("HasRulesManager", "Deleted")
  End Sub


End Class
